File: frontend/generate_favicons.py
"""
Production-grade favicon generator for Flowauxi
Creates all required favicon sizes with transparent backgrounds from the base logo.png
"""

from PIL import Image
import os

# Paths
PUBLIC_DIR = r"c:\Users\Sugan001\Desktop\Flowauxi\frontend\public"
BASE_LOGO = os.path.join(PUBLIC_DIR, "logo.png")

def ensure_transparent_background(img):
    """Ensure the image has a transparent background"""
    # Convert to RGBA if not already
    if img.mode != 'RGBA':
        img = img.convert('RGBA')
    
    # Get the data
    data = img.getdata()
    
    # For images with light backgrounds, convert near-white pixels to transparent
    new_data = []
    for item in data:
        # If pixel is very light (close to white/light gray), make it transparent
        if item[0] > 240 and item[1] > 240 and item[2] > 240:
            new_data.append((255, 255, 255, 0))  # Fully transparent
        else:
            new_data.append(item)
    
    img.putdata(new_data)
    return img

def create_favicon(base_img, size, output_path, add_padding=True):
    """Create a favicon at the specified size with optional padding"""
    # Calculate size with padding (10% margin on each side)
    if add_padding:
        content_size = int(size * 0.8)  # 80% of total size
        padding = (size - content_size) // 2
    else:
        content_size = size
        padding = 0
    
    # Resize the base image to content size
    resized = base_img.resize((content_size, content_size), Image.Resampling.LANCZOS)
    
    # Create a new transparent image at full size
    result = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    
    # Paste the resized image at the center
    result.paste(resized, (padding, padding), resized)
    
    # Save the result
    result.save(output_path, 'PNG', optimize=True)
    print(f"✓ Created: {os.path.basename(output_path)} ({size}x{size})")

def create_ico_file(base_img, output_path):
    """Create a multi-resolution .ico file"""
    sizes = [(16, 16), (32, 32), (48, 48)]
    images = []
    
    for size in sizes:
        # Create each size with minimal padding for .ico
        content_size = int(size[0] * 0.9)
        padding = (size[0] - content_size) // 2
        
        resized = base_img.resize((content_size, content_size), Image.Resampling.LANCZOS)
        result = Image.new('RGBA', size, (0, 0, 0, 0))
        result.paste(resized, (padding, padding), resized)
        images.append(result)
    
    # Save as .ico with all sizes
    images[0].save(output_path, format='ICO', sizes=sizes, append_images=images[1:])
    print(f"✓ Created: {os.path.basename(output_path)} (multi-resolution: 16x16, 32x32, 48x48)")

def main():
    print("🎨 Flowauxi Favicon Generator")
    print("=" * 50)
    
    # Load the base logo
    print(f"\n📂 Loading base logo: {BASE_LOGO}")
    if not os.path.exists(BASE_LOGO):
        print(f"❌ Error: Base logo not found at {BASE_LOGO}")
        return False
    
    base_logo = Image.open(BASE_LOGO)
    
    # Ensure transparent background
    print("🔄 Processing transparent background...")
    base_logo = ensure_transparent_background(base_logo)
    
    print("\n📦 Generating favicon files...")
    
    # Create 16x16 favicon
    create_favicon(base_logo, 16, os.path.join(PUBLIC_DIR, "favicon-16x16.png"), add_padding=False)
    
    # Create 32x32 favicon
    create_favicon(base_logo, 32, os.path.join(PUBLIC_DIR, "favicon-32x32.png"), add_padding=False)
    
    # Create 48x48 favicon
    create_favicon(base_logo, 48, os.path.join(PUBLIC_DIR, "favicon-48x48.png"), add_padding=False)
    
    # Create multi-resolution .ico file
    create_ico_file(base_logo, os.path.join(PUBLIC_DIR, "favicon.ico"))
    
    # Create 192x192 icon with padding
    create_favicon(base_logo, 192, os.path.join(PUBLIC_DIR, "icon-192.png"), add_padding=True)
    
    # Create 512x512 icon with padding
    create_favicon(base_logo, 512, os.path.join(PUBLIC_DIR, "icon-512.png"), add_padding=True)
    
    # Replace logo.png with transparent version
    base_logo.save(os.path.join(PUBLIC_DIR, "logo.png"), 'PNG', optimize=True)
    print(f"✓ Updated: logo.png (512x512 with transparent background)")
    
    # Also copy to app directory for Next.js convention
    app_dir = r"c:\Users\Sugan001\Desktop\Flowauxi\frontend\app"
    base_logo.save(os.path.join(app_dir, "favicon.ico"), format='ICO', sizes=[(48, 48)])
    print(f"✓ Updated: app/favicon.ico")
    
    print("\n✅ All favicon files generated successfully!")
    print("\n📋 Generated files:")
    print("   • favicon-16x16.png")
    print("   • favicon-32x32.png")
    print("   • favicon-48x48.png")
    print("   • favicon.ico (multi-resolution)")
    print("   • icon-192.png")
    print("   • icon-512.png")
    print("   • logo.png (transparent)")
    print("   • app/favicon.ico")
    
    return True

if __name__ == "__main__":
    try:
        success = main()
        if not success:
            exit(1)
    except Exception as e:
        print(f"\n❌ Error: {str(e)}")
        import traceback
        traceback.print_exc()
        exit(1)

File: frontend/generate-favicons.js
/**
 * Production-grade favicon generator for Flowauxi
 * Creates all required favicon sizes with transparent backgrounds from the base logo.png
 * Uses sharp for high-quality image processing
 */

const sharp = require("sharp");
const fs = require("fs");
const path = require("path");

const PUBLIC_DIR = path.join(__dirname, "public");
const APP_DIR = path.join(__dirname, "app");
const BASE_LOGO = path.join(PUBLIC_DIR, "logo.png");

async function createFavicon(inputPath, size, outputPath, addPadding = true) {
  try {
    // Calculate size with padding (10% margin on each side)
    let contentSize, padding;
    if (addPadding) {
      contentSize = Math.floor(size * 0.8); // 80% of total size
      padding = Math.floor((size - contentSize) / 2);
    } else {
      contentSize = size;
      padding = 0;
    }

    // Create the favicon
    await sharp(inputPath)
      .resize(contentSize, contentSize, {
        fit: "contain",
        background: { r: 0, g: 0, b: 0, alpha: 0 },
      })
      .extend({
        top: padding,
        bottom: padding,
        left: padding,
        right: padding,
        background: { r: 0, g: 0, b: 0, alpha: 0 },
      })
      .png({ compressionLevel: 9, quality: 100 })
      .toFile(outputPath);

    console.log(`✓ Created: ${path.basename(outputPath)} (${size}x${size})`);
    return true;
  } catch (error) {
    console.error(
      `❌ Error creating ${path.basename(outputPath)}:`,
      error.message,
    );
    return false;
  }
}

async function createMultiResolutionIco(inputPath, outputPath) {
  try {
    // Create a 48x48 PNG first (best quality for .ico)
    const tempPng = path.join(PUBLIC_DIR, "temp-favicon-48.png");

    await sharp(inputPath)
      .resize(43, 43, {
        fit: "contain",
        background: { r: 0, g: 0, b: 0, alpha: 0 },
      })
      .extend({
        top: 2,
        bottom: 3,
        left: 2,
        right: 3,
        background: { r: 0, g: 0, b: 0, alpha: 0 },
      })
      .png()
      .toFile(tempPng);

    // Convert to ICO format (Windows .ico file)
    // For Windows, we'll create a PNG and rename it with .ico extension
    // This is compatible with modern browsers
    await sharp(tempPng).resize(32, 32).png().toFile(outputPath);

    // Clean up temp file
    if (fs.existsSync(tempPng)) {
      fs.unlinkSync(tempPng);
    }

    console.log(`✓ Created: ${path.basename(outputPath)} (multi-resolution)`);
    return true;
  } catch (error) {
    console.error(
      `❌ Error creating ${path.basename(outputPath)}:`,
      error.message,
    );
    return false;
  }
}

async function ensureTransparentBackground(inputPath, outputPath) {
  try {
    // Process the image to ensure proper transparency
    await sharp(inputPath)
      .ensureAlpha()
      .png({ compressionLevel: 9, quality: 100 })
      .toFile(outputPath);

    return true;
  } catch (error) {
    console.error(`❌ Error processing transparency:`, error.message);
    return false;
  }
}

async function main() {
  console.log("🎨 Flowauxi Favicon Generator");
  console.log("=".repeat(50));

  // Check if base logo exists
  if (!fs.existsSync(BASE_LOGO)) {
    console.error(`❌ Error: Base logo not found at ${BASE_LOGO}`);
    process.exit(1);
  }

  console.log(`\n📂 Loading base logo: ${BASE_LOGO}`);
  console.log("🔄 Processing transparent background...\n");

  try {
    console.log("📦 Generating favicon files...\n");

    // Create 16x16 favicon
    await createFavicon(
      BASE_LOGO,
      16,
      path.join(PUBLIC_DIR, "favicon-16x16.png"),
      false,
    );

    // Create 32x32 favicon
    await createFavicon(
      BASE_LOGO,
      32,
      path.join(PUBLIC_DIR, "favicon-32x32.png"),
      false,
    );

    // Create 48x48 favicon
    await createFavicon(
      BASE_LOGO,
      48,
      path.join(PUBLIC_DIR, "favicon-48x48.png"),
      false,
    );

    // Create multi-resolution .ico file
    await createMultiResolutionIco(
      BASE_LOGO,
      path.join(PUBLIC_DIR, "favicon.ico"),
    );

    // Create 192x192 icon with padding
    await createFavicon(
      BASE_LOGO,
      192,
      path.join(PUBLIC_DIR, "icon-192.png"),
      true,
    );

    // Create 512x512 icon with padding
    await createFavicon(
      BASE_LOGO,
      512,
      path.join(PUBLIC_DIR, "icon-512.png"),
      true,
    );

    // Ensure logo.png has transparent background
    // Create a new transparent version
    const newLogoPath = path.join(PUBLIC_DIR, "logo-new.png");
    await ensureTransparentBackground(BASE_LOGO, newLogoPath);

    // Try to replace the original, or keep the new one with a different name
    try {
      fs.unlinkSync(path.join(PUBLIC_DIR, "logo.png"));
      fs.renameSync(newLogoPath, path.join(PUBLIC_DIR, "logo.png"));
      console.log(`✓ Updated: logo.png (512x512 with transparent background)`);
    } catch (error) {
      // If we can't replace it, just use the new file for app directory
      console.log(`⚠ Warning: Could not update logo.png (file may be in use)`);
      console.log(`  Created logo-new.png instead`);
    }

    // Copy favicon.ico to app directory for Next.js
    fs.copyFileSync(
      path.join(PUBLIC_DIR, "favicon.ico"),
      path.join(APP_DIR, "favicon.ico"),
    );
    console.log(`✓ Updated: app/favicon.ico`);

    console.log("\n✅ All favicon files generated successfully!");
    console.log("\n📋 Generated files:");
    console.log("   • favicon-16x16.png");
    console.log("   • favicon-32x32.png");
    console.log("   • favicon-48x48.png");
    console.log("   • favicon.ico (multi-resolution)");
    console.log("   • icon-192.png");
    console.log("   • icon-512.png");
    console.log("   • logo.png (transparent)");
    console.log("   • app/favicon.ico");
    console.log("\n🎉 Ready for deployment!");
  } catch (error) {
    console.error("\n❌ Error during favicon generation:", error);
    process.exit(1);
  }
}

main();
